File: CS/TGAPE_CS/Engine/Renderer/VertexAttributes.cs
using CSGL;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TGAPE_CS.Engine.Renderer
{
    public struct VaoInitData
    {
        public byte AttrbuteCount;
        public byte[] AttributeSize;
        public uint[] AttributeType;
        public byte[] AttributeNormal;
        public uint[] AttributeOffset;
    }
    public class VertexAttributes
    {
        private uint _vboBufferHandle;
        private uint _iVboBufferHandle;
        private uint _vaoBufferHandle;

        public uint IndiceCount;

        private void _createBuffer(byte[] data, uint additionalBufferSize)
        {
            OpenGL.glGenBuffers(1, ref _vboBufferHandle);
            OpenGL.glBindBuffer(OpenGL.GL_ARRAY_BUFFER, _vboBufferHandle);
            OpenGL.glBufferData(OpenGL.GL_ARRAY_BUFFER, (int)(data.Length + additionalBufferSize) * sizeof(byte), IntPtr.Zero, OpenGL.GL_STATIC_DRAW);
            unsafe
            {
                fixed (void* ptrData = data)
                    OpenGL.glBufferSubData(OpenGL.GL_ARRAY_BUFFER, 0, data.Length * sizeof(byte), (IntPtr)ptrData);
            }
            OpenGL.glBindBuffer(OpenGL.GL_ARRAY_BUFFER, 0);

            IndiceCount = 0;
        }
        private void _createBuffer(byte[] data, ushort[] indiceData)
        {
            OpenGL.glGenBuffers(1, ref _vboBufferHandle);
            OpenGL.glBindBuffer(OpenGL.GL_ARRAY_BUFFER, _vboBufferHandle);
            unsafe
            {
                fixed (void* ptrData = data)
                    OpenGL.glBufferData(OpenGL.GL_ARRAY_BUFFER, data.Length * sizeof(byte), (IntPtr)ptrData, OpenGL.GL_STATIC_DRAW);
            }
            OpenGL.glBindBuffer(OpenGL.GL_ARRAY_BUFFER, 0);

            IndiceCount = (uint)indiceData.Length;
            OpenGL.glGenBuffers(1, ref _iVboBufferHandle);
            OpenGL.glBindBuffer(OpenGL.GL_ELEMENT_ARRAY_BUFFER, _iVboBufferHandle);
            unsafe
            {
                fixed (void* ptrIData = indiceData)
                    OpenGL.glBufferData(OpenGL.GL_ELEMENT_ARRAY_BUFFER, indiceData.Length * sizeof(ushort), (IntPtr)ptrIData, OpenGL.GL_STATIC_DRAW);
            }
            OpenGL.glBindBuffer(OpenGL.GL_ELEMENT_ARRAY_BUFFER, 0);
        }
        private void _createBuffer(byte[] data, uint[] indiceData)
        {
            OpenGL.glGenBuffers(1, ref _vboBufferHandle);
            OpenGL.glBindBuffer(OpenGL.GL_ARRAY_BUFFER, _vboBufferHandle);
            unsafe
            {
                fixed (void* ptrData = data)
                    OpenGL.glBufferData(OpenGL.GL_ARRAY_BUFFER, data.Length * sizeof(byte), (IntPtr)ptrData, OpenGL.GL_STATIC_DRAW);
            }
            OpenGL.glBindBuffer(OpenGL.GL_ARRAY_BUFFER, 0);

            IndiceCount = (uint)indiceData.Length;
            OpenGL.glGenBuffers(1, ref _iVboBufferHandle);
            OpenGL.glBindBuffer(OpenGL.GL_ELEMENT_ARRAY_BUFFER, _iVboBufferHandle);
            unsafe
            {
                fixed (void* ptrIData = indiceData)
                    OpenGL.glBufferData(OpenGL.GL_ELEMENT_ARRAY_BUFFER, indiceData.Length * sizeof(uint), (IntPtr)ptrIData, OpenGL.GL_STATIC_DRAW);
            }
            OpenGL.glBindBuffer(OpenGL.GL_ELEMENT_ARRAY_BUFFER, 0);
        }
        private void _createVertexArray(VaoInitData vaoInitData)
        {
            OpenGL.glGenVertexArrays(1, ref _vaoBufferHandle);
            OpenGL.glBindVertexArray(_vaoBufferHandle);
            OpenGL.glBindBuffer(OpenGL.GL_ARRAY_BUFFER, _vboBufferHandle);
            OpenGL.glBindBuffer(OpenGL.GL_ELEMENT_ARRAY_BUFFER, _iVboBufferHandle);
            if(vaoInitData.AttrbuteCount > 16)
            {
                Console.WriteLine("Warning : attribute count is greater than 16");
                vaoInitData.AttrbuteCount = 16;
            }
            for(uint i = 0; i < vaoInitData.AttrbuteCount; i++)
            {
                OpenGL.glEnableVertexAttribArray(i);
                if(vaoInitData.AttributeType[i] == OpenGL.GL_FLOAT)
                {
                    OpenGL.glVertexAttribPointer(i, vaoInitData.AttributeSize[i], vaoInitData.AttributeType[i], vaoInitData.AttributeNormal[i], 0, (IntPtr)0 + (int)vaoInitData.AttributeOffset[i]);
                }
                else
                {
                    OpenGL.glVertexAttribIPointer(i, vaoInitData.AttributeSize[i], vaoInitData.AttributeType[i], 0, (IntPtr)0 + (int)vaoInitData.AttributeOffset[i]);
                }
            }
            OpenGL.glBindVertexArray(0);
        }

        public void Use()
        {
            OpenGL.glBindVertexArray(_vaoBufferHandle);
            OpenGL.glBindBuffer(OpenGL.GL_ARRAY_BUFFER, _vboBufferHandle);
            OpenGL.glBindBuffer(OpenGL.GL_ELEMENT_ARRAY_BUFFER, _iVboBufferHandle);
        }
        public void UpdateBuffer(byte[] data, uint offset, uint size)
        {
            OpenGL.glBindBuffer(OpenGL.GL_ARRAY_BUFFER, _vboBufferHandle);
            unsafe
            {
                fixed (void* ptrData = data)
                    OpenGL.glBufferSubData(OpenGL.GL_ARRAY_BUFFER, (int) offset, (int)size, (IntPtr)ptrData);
            }
            OpenGL.glBindBuffer(OpenGL.GL_ARRAY_BUFFER, 0);

        }
    }
}

File: TGAPEngine/src/VertexAttributes.cpp
#include "VertexAttributes.h"
namespace engine{
VertexAttributes::VertexAttributes()
{
    //ctor
}
VertexAttributes::VertexAttributes(std::vector<char> data, std::vector<unsigned short> iData, VaoInitData initData)
{
    createBuffer(data,iData);
    createVertexArray(initData);

}
VertexAttributes::VertexAttributes(std::vector<char> data, std::vector<unsigned int> iData, VaoInitData initData)
{
    createBuffer(data,iData);
    createVertexArray(initData);

}

VertexAttributes::VertexAttributes(std::vector<char> data, VaoInitData initData, unsigned int additionalBufferSize){
	createBuffer(data, additionalBufferSize);
	createVertexArray(initData);

}

void VertexAttributes::createBuffer(std::vector<char> data, unsigned int additionalBufferSize)
{
	glGenBuffers(1, &this->vboBuffer);
	glBindBuffer(GL_ARRAY_BUFFER, this->vboBuffer);
	glBufferData(GL_ARRAY_BUFFER, data.size() + additionalBufferSize *sizeof(char), NULL, GL_STATIC_DRAW);
	glBufferSubData(GL_ARRAY_BUFFER, 0, data.size() * sizeof(char), &data.front());
	glBindBuffer(GL_ARRAY_BUFFER, 0);

	this->indiceCount = 0;
	
}

VertexAttributes::~VertexAttributes()
{
    //dtor
}
void VertexAttributes::createBuffer(std::vector<char> data, std::vector<unsigned short> iData)
{
    indiceType = GL_UNSIGNED_SHORT;
    glGenBuffers(1,&this->vboBuffer);
    glBindBuffer(GL_ARRAY_BUFFER,this->vboBuffer);
    glBufferData(GL_ARRAY_BUFFER,data.size()*sizeof(char),&data.front(),GL_STATIC_DRAW);
    glBindBuffer(GL_ARRAY_BUFFER,0);

    this->indiceCount = iData.size();
    glGenBuffers(1,&this->iVboBuffer);
    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER,this->iVboBuffer);
    glBufferData(GL_ELEMENT_ARRAY_BUFFER,iData.size()*sizeof(short),&iData.front(),GL_STATIC_DRAW);
    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER,0);

}
void VertexAttributes::createBuffer(std::vector<char> data, std::vector<unsigned int> iData)
{
    indiceType = GL_UNSIGNED_INT;
    glGenBuffers(1,&this->vboBuffer);
    glBindBuffer(GL_ARRAY_BUFFER,this->vboBuffer);
    glBufferData(GL_ARRAY_BUFFER,data.size()*sizeof(char),&data.front(),GL_STATIC_DRAW);
    glBindBuffer(GL_ARRAY_BUFFER,0);

    this->indiceCount = iData.size();
    glGenBuffers(1,&this->iVboBuffer);
    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER,this->iVboBuffer);
    glBufferData(GL_ELEMENT_ARRAY_BUFFER,iData.size()*sizeof(int),&iData.front(),GL_STATIC_DRAW);
    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER,0);

}
void VertexAttributes::createVertexArray(VaoInitData initData)
{
    glGenVertexArrays(1,&this->vaoBuffer);
    glBindVertexArray(this->vaoBuffer);
    glBindBuffer(GL_ARRAY_BUFFER,this->vboBuffer);
    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER,this->iVboBuffer);
    if(initData.attrNumber>16)
    {
        std::cerr<<"Warning : attribute number is greater than 16"<<std::endl;
        initData.attrNumber=16;
    }
    int i;
    for(i=0;i<initData.attrNumber;++i){
        glEnableVertexAttribArray(i);
        if(initData.attrType[i] == GL_FLOAT){
            glVertexAttribPointer(i,initData.attrSize[i],initData.attrType[i],initData.attrNormBool[i],0,(void*)initData.attrOffset[i]);
        }else{
            glVertexAttribIPointer(i,initData.attrSize[i],initData.attrType[i],0,(void*)initData.attrOffset[i]);
        }
    }
    glBindVertexArray(0);
}
void VertexAttributes::use()
{
    glBindVertexArray(this->vaoBuffer);
    glBindBuffer(GL_ARRAY_BUFFER,this->vboBuffer);
    glBindBuffer(GL_ELEMENT_ARRAY_BUFFER,this->iVboBuffer);


}

void VertexAttributes::updateBuffer(char * data, unsigned int offset, unsigned int size){
	glBindBuffer(GL_ARRAY_BUFFER, this->vboBuffer);
	glBufferSubData(GL_ARRAY_BUFFER, offset, size, data);

	glBindBuffer(GL_ARRAY_BUFFER, 0);
}

}
